doc: clarify dep spec into (#117001)
NixOS/nixpkgs: doc/stdenv/stdenv.chapter.md

## Specifying dependencies {#ssec-stdenv-dependencies}

> While _derivations_ collect their dependencies by references to store paths of other _derivations_, `mkDerivation`, furthermore deals with _package_ depdencies along four axes: build, host & target platform and wether they are propagated. Details can be found in <xref linkend="chap-cross" />, but, for quick reminder, the machine you are building on is _build_, the machine that you are building for _host_, and the machine that the compiler will produce code for _target_. If build, host, and target are all the same, this is called a _native_. If build and host are the same but target is different, this is called a _cross_. If host and target are the same, but build is different, you are using a _cross-compiler_ to build a _native_ for a different system, a _cross-built native_, because target is the machine that the compiler will produce code for, while running on build.

Author: ```suggestion
While _derivations_ collect their dependencies by references to store paths of other _derivations_, `mkDerivation`, furthermore deals with _package_ depdencies along four axes: build, host & target platform and wether they are propagated. Details can be found in <xref linkend="chap-cross" />, but, for quick reminder, 
- the machine you are building on is _build_, 
- the machine that you are building for _host_, 
- and the machine that the compiler will produce code for _target_. 

Accordingly:
- If build, host, and target are all the same, this is called a _native_. 
- If build and host are the same but target is different, this is called a _cross_.
- If host and target are the same, but build is different, you are using a _cross-compiler_ to build a _native_ for a different system, a _cross-built native_, because target is the machine that the compiler will produce code for, while running on build.
```

But even if one is not cross compiling, the platforms imply whether a dependency is needed at run-time or build-time.

The extension of `PATH` is accomplished by simply adding `dep/bin`, if present. Understandibly, this is only done for dependencies who actually also run ("host") on the new derivation's build platform.

> While Nix perfectly manages transitive dependencies, a _propagated_ depdency is one that (forward-)inserts itself into transitive downstream packages as _direct_ dependency. Therefore eventual downstream's setup-hooks and tooling also affect it. It is important to note, that propagated dependencies, since they are "fast-forwarded" to the downstream derviation, a _cross_ dependency might actually end up as a _native_ dependency.

Author: ```suggestion
While Nix perfectly manages transitive dependencies, a _propagated_ depdency is one that (forward-)inserts itself into transitive downstream packages as _direct_ dependency. Therefore eventual downstream's setup-hooks and tooling also affect it. It is important to note for propagated dependencies, since they are "fast-forwarded" to the downstream derviation, that a _cross_ dependency might actually end up as a _native_ dependency.
```

> In order to propagate/flatten a dependency graph of _propagated_ dependencies, we describe the their platform-offset to their immediate depending propagated dependency by assigning two integers `h` (host-offset) & `t` (target-offset). We then traverse propagated inputs and fold their offsets as follows:

Author: ```suggestion
In order to propagate/flatten a dependency graph of _propagated_ dependencies, we describe their platform-offset to their immediate depending propagated dependency by assigning two integers `h` (host-offset) & `t` (target-offset). We then traverse propagated inputs and fold their offsets as follows:
```

let f(h, h + 1, i) = i + h
```

Because of the bounds checks, the uncommon cases are `h = t` and `h + 2 = t`. In the former case, the motivation for `mapOffset` is that since its host and target platforms are the same, no transitive dependency of it should be able to “discover” an offset greater than its reduced target offsets. `mapOffset` effectively “squashes” all its transitive dependencies’ offsets so that none will ever be greater than the target offset of the original `h = t` package. In the other case, `h + 1` is skipped over between the host and target offsets. Instead of squashing the offsets, we need to “rip” them apart so no transitive dependencies’ offset is that one.

> _Remember, the big picture is that propagation must not introduce transitive dependencies involving platforms the depending derivation doesn't know. The offset bounds checking and definition of <function>mapOffset</function> together ensure that this is the case. Discovering a new offset is discovering a new platform, and since those platforms weren't in the derivation specification, they cannot be relevant. From a capability perspective, we can imagine that the host and target platforms of a package are the capabilities a package requires, and the depending package must provide the capability to the dependency._

Author: ```suggestion
_Remember, the big picture is that propagation must not introduce transitive dependencies involving platforms the depending derivation doesn't know. The offset bounds implemented in `mapOffset` make sure of that. Discovering a new offset is discovering a new platform, and since those platforms weren't in the derivation specification, they cannot be relevant. From a capability perspective, we can imagine that the host and target platforms of a package are the capabilities a package requires, and the depending package must provide the capability to the dependency._
```

For more in-depth information on these hardening flags and hardening in general, refer to the [Debian Wiki](https://wiki.debian.org/Hardening), [Ubuntu Wiki](https://wiki.ubuntu.com/Security/Features), [Gentoo Wiki](https://wiki.gentoo.org/wiki/Project:Hardened), and the [Arch Wiki](https://wiki.archlinux.org/index.php/DeveloperWiki:Security).

> [^footnote-stdenv-native-dependencies-in-path]: Currently, this means for native builds all dependencies are put on the `PATH`. But in the future that may not be the case for sake of matching cross: the platforms would be assumed to be unique for native and cross builds alike, so only the `depsBuild*` and `nativeBuildInputs` would be added to the `PATH`.

Author: ```suggestion
```

> [^footnote-stdenv-ignored-build-platform]: The build platform is ignored because it is a mere implementation detail of the package satisfying the dependency: As a general programming principle, dependencies are always *specified* as interfaces, not concrete implementation.

Author: ```suggestion
```
